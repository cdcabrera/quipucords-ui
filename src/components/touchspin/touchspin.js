import React from 'react';
import PropTypes from 'prop-types';
import { Form, Icon, InputGroup, Button } from 'patternfly-react';
import helpers from '../../common/helpers';

class TouchSpin extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      displayValue: Number.parseInt(props.value, 10),
      maxValue: Number.parseInt(props.maxValue, 10),
      minValue: Number.parseInt(props.minValue, 10),
      value: Number.parseInt(props.value, 10)
    };

    this.timer = null;
  }

  onFocusMax = () => {
    const { maxValue, value } = this.state;

    const setTimer = (time = 500) => {
      clearTimeout(this.timer);

      this.timer = setTimeout(() => {
        this.onMax();

        if (value < maxValue) {
          setTimer(0);
        }
      }, time);
    };

    setTimer();
  };

  onFocusMin = () => {
    const { minValue, value } = this.state;

    const setTimer = (time = 500) => {
      clearTimeout(this.timer);

      this.timer = setTimeout(() => {
        this.onMin();

        if (value > minValue) {
          setTimer(0);
        }
      }, time);
    };

    setTimer();
  };

  onBlur = () => {
    clearTimeout(this.timer);
  };

  onMin = () => {
    const { maxValue, minValue, value } = this.state;
    let updatedValue = value - 1 > minValue ? value - 1 : minValue;

    if (updatedValue > maxValue) {
      updatedValue = maxValue;
    }

    this.onUpdateValue({ target: { value: updatedValue } });
  };

  onMax = () => {
    const { maxValue, minValue, value } = this.state;
    let updatedValue = value + 1 < maxValue ? value + 1 : maxValue;

    if (updatedValue < minValue) {
      updatedValue = minValue;
    }

    this.onUpdateValue({ target: { value: updatedValue } });
  };

  onUpdateValue = event => {
    const { maxValue, minValue } = this.state;
    const { name, onChange } = this.props;
    const { value } = event.target;

    let parsedValue = Number.parseInt(value, 10);
    parsedValue = Number.isNaN(parsedValue) ? minValue : parsedValue;

    const maxDisabled = parsedValue > maxValue;
    const minDisabled = parsedValue < minValue;

    clearTimeout(this.timer);

    this.setState({
      displayValue: value,
      value: parsedValue,
      maxDisabled,
      minDisabled
    });

    const mockTarget = {
      ...event.target,
      name,
      value: parsedValue
    };

    const mockEvent = {
      ...mockTarget,
      target: { ...mockTarget },
      currentTarget: { ...mockTarget },
      persist: helpers.noop
    };

    return onChange(mockEvent);
  };

  render() {
    const { displayValue, maxDisabled, minDisabled } = this.state;
    const { className, labelMax, labelMaxDescription, labelMin, labelMinDescription, name } = this.props;

    return (
      <InputGroup className={`cloudmeter-touchspin ${className}`}>
        <InputGroup.Button>
          <Button
            onClick={this.onMin}
            onMouseDown={this.onFocusMin}
            onMouseUp={this.onBlur}
            onMouseLeave={this.onBlur}
            className="cloudmeter-touchspin-min-button form-control"
            aria-hidden
            tabIndex={-1}
            title={labelMinDescription}
            disabled={minDisabled}
          >
            {labelMin}
          </Button>
        </InputGroup.Button>
        <Form.FormControl
          type="text"
          name={name}
          value={displayValue}
          className="cloudmeter-touchspin-input"
          onChange={this.onUpdateValue}
        />
        <InputGroup.Button>
          <Button
            onClick={this.onMax}
            onMouseDown={this.onFocusMax}
            onMouseUp={this.onBlur}
            onMouseLeave={this.onBlur}
            className="cloudmeter-touchspin-max-button form-control"
            aria-hidden
            tabIndex={-1}
            title={labelMaxDescription}
            disabled={maxDisabled}
          >
            {labelMax}
          </Button>
        </InputGroup.Button>
      </InputGroup>
    );
  }
}

TouchSpin.propTypes = {
  className: PropTypes.string,
  labelMax: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  labelMaxDescription: PropTypes.string,
  labelMin: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  labelMinDescription: PropTypes.string,
  maxValue: PropTypes.number,
  minValue: PropTypes.number,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

TouchSpin.defaultProps = {
  className: '',
  labelMax: <Icon type="fa" name="plus" />,
  labelMaxDescription: 'Increase number input',
  labelMin: <Icon type="fa" name="minus" />,
  labelMinDescription: 'Decrease number input',
  maxValue: 100,
  minValue: 0,
  onChange: helpers.noop,
  value: 50
};

export { TouchSpin as default, TouchSpin };
