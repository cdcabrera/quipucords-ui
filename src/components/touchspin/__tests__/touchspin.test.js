import React from 'react';
import { mount } from 'enzyme';
import { TextInput } from '@patternfly/react-core';
import { TouchSpin } from '../touchspin';

describe('TouchSpin Component', () => {
  it('should render a basic component', () => {
    const props = {
      name: 'lorem'
    };

    const component = mount(<TouchSpin {...props} />);

    const componentInstance = component.instance();
    expect(componentInstance.props).toMatchSnapshot('initial props');
    expect(componentInstance.state).toMatchSnapshot('initial state');

    expect(component.render()).toMatchSnapshot('basic render');
  });

  it('should handle onchange events', () => {
    const props = {
      name: 'lorem',
      value: 1,
      labelMaxDescription: 'testing max',
      labelMinDescription: 'testing min'
    };

    const component = mount(<TouchSpin {...props} />);

    component.find(`button[aria-label="${props.labelMaxDescription}"]`).simulate('click');
    expect(component.find(TextInput).props()?.value).toBe(2);

    component.find(`button[aria-label="${props.labelMinDescription}"]`).simulate('click');
    expect(component.find(TextInput).props()?.value).toBe(1);
  });
});
