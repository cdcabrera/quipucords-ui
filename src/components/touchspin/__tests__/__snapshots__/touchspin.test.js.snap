// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TouchSpin Component should render a basic component: basic render 1`] = `
<span
  class="cloudmeter-touchspin  input-group"
>
  <span
    class="input-group-btn"
  >
    <button
      aria-hidden="true"
      class="cloudmeter-touchspin-min-button form-control btn btn-default"
      tabindex="-1"
      title="Decrease number input"
      type="button"
    >
      <span
        aria-hidden="true"
        class="fa fa-minus"
      />
    </button>
  </span>
  <input
    class="cloudmeter-touchspin-input form-control"
    name="lorem"
    type="text"
    value="50"
  />
  <span
    class="input-group-btn"
  >
    <button
      aria-hidden="true"
      class="cloudmeter-touchspin-max-button form-control btn btn-default"
      tabindex="-1"
      title="Increase number input"
      type="button"
    >
      <span
        aria-hidden="true"
        class="fa fa-plus"
      />
    </button>
  </span>
</span>
`;

exports[`TouchSpin Component should render a basic component: initial props 1`] = `
Object {
  "className": "",
  "labelMax": <Icon
    name="plus"
    type="fa"
  />,
  "labelMaxDescription": "Increase number input",
  "labelMin": <Icon
    name="minus"
    type="fa"
  />,
  "labelMinDescription": "Decrease number input",
  "maxValue": 100,
  "minValue": 0,
  "name": "lorem",
  "onChange": [Function],
  "value": 50,
}
`;

exports[`TouchSpin Component should render a basic component: initial state 1`] = `
Object {
  "displayValue": 50,
  "maxValue": 100,
  "minValue": 0,
  "value": 50,
}
`;
